client/services/CodeStorageService.js

```diff
@@ -0,0 +1,42 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+ * A service that saves student code to the browser's localStorage.
+ */
+tie.factory('CodeStorageService', [
+  function() {
+    var codeStorageService = {};
+    
+    codeStorageService.getLocalStorageHashKey = function(questionId, language) {
+      return "[" + questionId + ":" + language + "]";
```

Comment:
For this requirement: Would be also nice to add karma tests in assets/tests/QuestionSchemaValidationServiceSpec.js to force the question id to contain only alphanumeric chars + digits (and, in particular, no colons). If we ever want to decode keys this might help with preventing ambiguity in the decoding.

I found it a little bit difficult to accomplish since there is a test case: validateAllQuestions, which will test against all functions with question as parameter. However, I cannot get questionId from question directly. I can choose to put the test in another place or change the way I compose the key. Please let me know.